Sources/Storage/Disk/DiskStorage.swift

```diff
@@ -0,0 +1,76 @@
+//
+//  DiskStorage.swift
+//  StatsdClient-iOS
+//
+//  Created by Nghia Tran on 10/8/17.
+//  Copyright © 2017 StatsdClient. All rights reserved.
+//
+
+import Foundation
+
+final class DiskStorage<Element: Serializable>: Storage {
+
+    typealias Key = String
+    typealias Item = Element
+
+    // MARK: - Variable
+    private let handler: PersistentHandler
+    private let queue = DispatchQueue(label: "StatsD_DiskStorage", qos: .default, attributes: .concurrent)
+
+    // Count
+    var count: Int {
+        return queue.syncWithReturnedValue { self.handler.getTotal() }
+    }
+
+    // MARK: - Init
+    init(handler: PersistentHandler) {
+        self.handler = handler
+    }
+
+    // Default
+    init?(config: DiskConfigurable) {
+
+        // Try to create default DiskPersistentHandler
+        let handler = try? DiskPersistentHandler(config: config)
+
+        // Make sure we could create proper DiskHandler with config
+        guard let disk = handler else {return nil }
+
+        // Save
+        self.handler = disk
+    }
+
+    // MARK: - Public
+    func item(forKey key: Key) -> Element?  {
+        return queue.syncWithReturnedValue {
+            try? handler.get(key: key, type: Element.self)
+        }
+    }
+
+    func set(item: Element, forKey key: Key) {
+      queue.async(flags: .barrier) { [unowned self] in
+
+            // Don't handle Throws error here
+            try? self.handler.write(item, key: key, attribute: nil)
```

Comment:
Would it be better to crash or handle the error? This will just silently fail to write to the cache.

Reply:
yup, I agree with you. At that time I can't throw an exception from this async Block, so I just put it temporarily. Let me think an alternative approach to handle it